### ISSUE TITLE
clang hits "Illegal instruction: 4" after "no matching function for call" on llvm-reduce 

### ISSUE THREAD

**Issue by arsenm** [llvm/llvm-project#58576]
Ran into this crash when working on llvm-reduce. Hit with Apple clang, also reproduces with upstream commit ddb6c2b301f60c8d5a80d984554b93590cc4bc76

[Clang crash.zip](https://github.com/llvm/llvm-project/files/9855313/Clang.crash.zip)

**Comment by Endilll** [comment]
Manual reduction lead to multiple similar crashes with multiple reproducers. Apparently this code brings Clang in a quite fragile state. We issue `stack-exhausted` diagnostic, but only for some cases.  Here is a non-exhaustive list of cases where we don't warn the user about stack exhaustion:
https://godbolt.org/z/T1vxEnfao
https://godbolt.org/z/r3PP768Y3
https://godbolt.org/z/4G31fTrxr
https://godbolt.org/z/ezxe6df1q
https://godbolt.org/z/5hcv9q8zd
As an example, here's the first test case:
```cpp
template <class _Tp>
struct remove_reference {
  typedef _Tp type;
};

template <class _Tp>
constexpr _Tp && move(_Tp &&);

namespace std {
template <class _Tp>
constexpr _Tp && forward(typename remove_reference<_Tp>::type &);
}

template <class>
class __bind {};

template <class _Fp>
__bind<_Fp> bind(_Fp &&);

namespace llvm {
class ThreadPool {
  template <typename Function>
  auto async(Function &&F) {
    auto Task = bind(std::forward<Function>(F));
    return async(Task);
  }
};

void llvm::runDeltaPass() {
  ThreadPool ChunkThreadPool;
  ChunkThreadPool.async([] {});
}
```
Clang 16 crash (don't have 17 with assertions at the moment):
```
pp.ii:29:12: warning: extra qualification on member 'runDeltaPass' [-Wextra-qualification]
void llvm::runDeltaPass() {
           ^
pp.ii:29:12: error: out-of-line definition of 'runDeltaPass' does not match any declaration in namespace 'llvm'
void llvm::runDeltaPass() {
           ^~~~~~~~~~~~
pp.ii:31:19: error: 'async' is a private member of 'llvm::ThreadPool'
  ChunkThreadPool.async([] {});
  ~~~~~~~~~~~~~~~~^~~~~
pp.ii:23:8: note: implicitly declared private here
  auto async(Function &&F) {
       ^
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-assertions-16.0.0/bin/clang -cc1 -triple x86_64-apple-macosx12.0.0 -emit-obj -std=c++17 -fno-rtti -x c++ pp.ii
1.	pp.ii:31:30: current parser token ')'
2.	pp.ii:20:1: parsing namespace 'llvm'
3.	pp.ii:29:27: parsing function body 'llvm::runDeltaPass'
4.	pp.ii:29:27: in compound statement ('{}')
5.	pp.ii:23:8: instantiating function definition 'llvm::ThreadPool::async<(lambda at pp.ii:31:25)>'
6.	pp.ii:23:8: instantiating function definition 'llvm::ThreadPool::async<__bind<(lambda at pp.ii:31:25)> &>'
7.	pp.ii:23:8: instantiating function definition 'llvm::ThreadPool::async<__bind<__bind<(lambda at pp.ii:31:25)> &> &>'
8.	pp.ii:23:8: instantiating function definition 'llvm::ThreadPool::async<__bind<__bind<__bind<(lambda at pp.ii:31:25)> &> &> &>'
9.	pp.ii:23:8: instantiating function definition 'llvm::ThreadPool::async<__bind<__bind<__bind<__bind<(lambda at pp.ii:31:25)> &> &> &> &>'
10.	pp.ii:23:8: instantiating function definition 'llvm::ThreadPool::async<__bind<__bind<__bind<__bind<__bind<(lambda at pp.ii:31:25)> &> &> &> &> &>'
11.	pp.ii:23:8: instantiating function definition 'llvm::ThreadPool::async<__bind<__bind<__bind<__bind<__bind<__bind<(lambda at pp.ii:31:25)> &> &> &> &> &> &>'
12.	pp.ii:23:8: instantiating function definition 'llvm::ThreadPool::async<__bind<__bind<__bind<__bind<__bind<__bind<__bind<(lambda at pp.ii:31:25)> &> &> &> &> &> &> &>'
[going on with the recursion]
348.	pp.ii:23:8: instantiating function definition 'llvm::ThreadPool::async<malloc(): unaligned tcache chunk detected
```
